### PR TITLE
Implements `Reader` monad.

### PR DESCRIPTION
**Pull request by drupol**
This PR:

- [x] Implements the `Reader` Monad.
- [ ] Has no Psalm errors

Keep in mind that I'm doing this to learn monads, it might contains inconsistencies - please bear with me.

### CODE REVIEW

File: src/Brand/ReaderBrand.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+declare(strict_types=1);
+
+namespace Marcosh\LamPHPda\Brand;
+
+use Marcosh\LamPHPda\Reader;
+
+/**
+ * @implements Brand<Reader>
+ */
```
Comment (marcosh) [inline src/Brand/ReaderBrand.php:11]: even though the class is empty, I'd use `@psalm-immutable` anyway

File: src/Instances/Reader/ReaderApplicative.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * @psalm-immutable
+ */
+final class ReaderApplicative implements Applicative
+{
+    /**
+     * @template A
+     * @template B
+     * @template E
+     * @param callable(A): B $f
+     * @param HK1<ReaderBrand, A> $a
```
Comment (marcosh) [inline src/Instances/Reader/ReaderApplicative.php:24]: here, and in other similar places, you could use `ReaderBrand<E>`

File: src/Instances/Reader/ReaderApply.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        $readerF = Reader::fromBrand($f);
+        $readerA = Reader::fromBrand($a);
+
+        return Reader::reader(
+            /**
+             * @param E $env
+             * @return B
+             */
+            static fn($env) => ($readerA->runReader($env))($readerF->runReader($env))
```
Comment (marcosh) [inline src/Instances/Reader/ReaderApply.php:54]: here the order of `$readerF` and `$readerA` is wrong. Like this you're trying to apply a value to a function

File: src/Reader.php [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+use Marcosh\LamPHPda\Typeclass\Apply;
+use Marcosh\LamPHPda\Typeclass\DefaultInstance\DefaultMonad;
+use Marcosh\LamPHPda\Typeclass\Functor;
+use Marcosh\LamPHPda\Typeclass\Monad;
+use Marcosh\LamPHPda\Brand\ReaderBrand;
+
+/**
+ * @template E
+ * @template A
+ * @implements DefaultMonad<ReaderBrand, A>
```
Comment (marcosh) [inline src/Reader.php:22]: add a newline here

File: src/Reader.php [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+     * @psalm-pure
+     */
+    public static function of($b): Reader
+    {
+        return new self(static fn () => $b);
+    }
+
+    /**
+     * @template B
+     * @param HK1<ReaderBrand, B> $b
```
Comment (marcosh) [inline src/Reader.php:69]: also here it'd be useful to use `ReaderBrand<E>`, to that the information about the environment type could be propagated

File: src/Reader.php [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+     * @psalm-mutation-free
+     */
+    private function __construct(callable $f)
+    {
+        $this->action = $f;
+    }
+
+    /**
+     * @template B
+     * @param callable(E): B $f
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Reader.php:45]: It's already defined in the class header. Should we also redefine it here?

File: src/Reader.php [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    /**
+     * @template B
+     * @param callable(E): B $b
+     * @return Reader<E, B>
+     *
+     * @psalm-pure
+     */
+    public static function of($b): Reader
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Reader.php]: fixed

File: src/Reader.php [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        return self::fromBrand($functor->map($f, $this));
+    }
+
+    /**
+     * @template B
+     * @param callable(A): B $f
+     * @return Reader<E, B>
+     *
+     * @psalm-pure
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Reader.php]: fixed.

File: src/Reader.php [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+     * @psalm-mutation-free
+     */
+    public function iapply(Apply $apply, HK1 $f): Reader
+    {
+        return self::fromBrand($apply->apply($f, $this));
+    }
+
+    /**
+     * @template B
+     * @param Apply<ReaderBrand, callable(A): B> $f
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Reader.php]: Fixed.

File: src/Reader.php [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    public function apply(HK1 $f): Reader
+    {
+        return $this->iapply(new ReaderApply(), $f);
+    }
+
+    /**
+     * @template B
+     * @param Applicative<ReaderBrand> $applicative
+     * @param B $a
+     * @return Reader<E, B>
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Reader.php:134]: Fixed.

File: src/Reader.php [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    public static function pure($a): self
+    {
+        return self::ipure(new ReaderApplicative(), $a);
+    }
+
+    /**
+     * @template B
+     * @param Monad<ReaderBrand> $monad
+     * @param callable(A): HK1<ReaderBrand, B> $f
+     * @return Reader<E, B>
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Reader.php:159]: Fixed.

File: src/Brand/ReaderBrand.php [hunk 12]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Marcosh\LamPHPda\Brand;
+
+use Marcosh\LamPHPda\Reader;
+
+/**
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Brand/ReaderBrand.php:9]: Fixed.

File: src/Instances/Reader/ReaderApply.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+     * @return Reader<E, B>
+     *
+     * @psalm-pure
+     */
+    public function apply(HK1 $f, HK1 $a): Reader
+    {
+        $readerF = Reader::fromBrand($f);
+        $readerA = Reader::fromBrand($a);
+
+        return Reader::reader(static fn($env) => ($readerA->runReader($env))($readerF->runReader($env)));
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Instances/Reader/ReaderApply.php]: I did it, but then I'm unable to fix the tests properly. Could you have a look?

File: spec/ReaderSpec.php [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+declare(strict_types=1);
+
+use Marcosh\LamPHPda\Reader;
+
+describe('Reader', function () {
+    it('apply', function () {
+        $addTwo = static fn (int $x): int => $x + 2;
+
+        $result = Reader::of($addTwo)->apply(Reader::of(10))->runReader(50);
```
[the remark this answers is not in the document]
Comment (drupol) [reply spec/ReaderSpec.php]: Fixed.

File: src/Brand/ReaderBrand.php (hunk 1 above)
Comment (drupol) [reply src/Brand/ReaderBrand.php:15]: Fixed.

File: src/Reader.php (hunk 4 above)
Comment (drupol) [reply src/Reader.php]: Fixed.

File: src/Reader.php (hunk 5 above)
Comment (drupol) [reply src/Reader.php]: Fixed.

File: src/Reader.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcosh) [reply src/Reader.php:46]: sorry, I made a typo. Since the method is static, I guess we need to use `@template F`, `callable(F): B` and `Reader<F, B>`

**Comment by marcosh** [comment]
I fixed this and merged it manually. Thanks a lot @drupol , great job!

**Pull request closed without merging**